include/deal.II/numerics/fe_field_convolution_function.h

```diff
@@ -0,0 +1,280 @@
+   * Functions::CutOffFunctionC1<dim> kernel;
+   * const double radius = 2*h;
+   *
+   * // Now build the convolution of the first function, projected on the
+   * // second domain
+   * FEFieldConvolutionFunction<dim> fe_function_1 (dh_1, solution_1,
+   *                                                kernel, radius);
+   * VectorTools::project (dh_2, constraints_2, quad,
+   *                       fe_function_1, solution_2);
+   *
+   * // Or interpolated...
+   * Vector<double> solution_3;
+   * VectorTools::interpolate (dh_2, fe_function_1, solution_3);
+   *
+   * @endcode
+   *
+   * FEFieldConvolutionFunction is designed to be an easy way to get the results
+   * of your computations across different, possibly non matching, grids. No
+   * knowledge of the location of the points is assumed in this class.
+   *
+   * <h3>Using FEFieldConvolutionFunction with
+   * parallel::distributed::Triangulation</h3>
+   *
+   * When using this class with a parallel distributed triangulation object
+   * and evaluating the solution at a particular point, not every processor
+   * will own the cell at which the solution is evaluated. Rather, it may be
+   * that the cell in which this point is found is in fact a ghost or
+   * artificial cell (see
+   * @ref GlossArtificialCell
+   * and
+   * @ref GlossGhostCell).
+   * The solution may be evaluated on non owned cells, but there the solution
+   * will vanish at a distance greater than `radius` from the locally owned
+   * domain. If you want to have a consistent solution on distributed
+   * Triangulation objects, you have to make sure you sum the results from all
+   * processors, i.e., using Utilities::MPI::sum(). In this way you are
+   * guaranteed that the actual convolution is performed with a non-zero field
+   * on only one processor, and the sum of the results across all processors
+   * will produce the same value you'd get with a serial Triangulation.
+   *
+   * @ingroup functions
+   * @author Luca Heltai, 2019
+   */
+  template <int dim,
+            int spacedim            = dim,
+            typename DoFHandlerType = DoFHandler<dim, spacedim>,
+            typename VectorType     = Vector<double>>
+  class FEFieldConvolutionFunction
+    : public Function<spacedim, typename VectorType::value_type>
+  {
+  public:
+    /**
+     * Construct a vector function. A smart pointers is stored to the dof
+     * handler and to the cache object, so you have to make sure they remain
+     * valid for the entire lifetime of this object. The number of
+     * components of this functions is equal to the number of components of the
+     * finite element object.
+     */
+    FEFieldConvolutionFunction(const DoFHandlerType &                 dh,
+                               const GridTools::Cache<dim, spacedim> &cache,
+                               const VectorType &data_vector,
+                               Functions::CutOffFunctionBase<spacedim> &kernel,
```

Comment:
Why is this not `const`?

Reply:
Internally we need to call the `set_center()` of `kernel`, which is non-const.

Reply:
We need to call `kernel.set_center()` for values and gradients evaluations, and that is a non-const method of kernel.